lib/vetext/service.rb

```diff
@@ -0,0 +1,140 @@
+    #
+    # @return Hash           response object, which includes endpoint_sid
+    #
+    def register(app_name, device_token, icn, os_info, device_name = nil)
+      app_sid = get_app_sid(app_name)
+      response = perform(
+        :put,
+        REGISTER_PATH, {
+          appSid: app_sid,
+          token: device_token,
+          icn: icn,
+          os: os_info[:name],
+          osVersion: os_info[:version],
+          deviceName: device_name || os_info[:name]
+        }
+      )
+      raise_if_response_error(response.body)
+      response.body
+    rescue Common::Client::Errors::ClientError => e
+      remap_error(e)
+    end
+
+    # Get preferences for a given user/device.
+    #
+    # @endpoint_sid String    the registration id as returned from `register`
+    #
+    # @return Hash            response object
+    #
+    def get_preferences(endpoint_sid)
+      response = perform(
+        :get,
+        "#{PREFERENCES_PATH}/#{endpoint_sid}",
+        nil
+      )
+      response.body
+    rescue Common::Client::Errors::ClientError => e
+      remap_error(e)
+    end
+
+    # Set a single preference for a given user/device.
+    #
+    # @endpoint_sid String          the registration id as returned from `register`
+    # @preference_id String         the preference type identifier
+    # @receive_preference boolean   true: user wishes to receive this type of push notification
+    #
+    # @return Hash                  response object
+    #
+    def set_preference(endpoint_sid, preference_id, receive_preference)
+      response = perform(
+        :put,
+        PREFERENCES_PATH, {
+          endpointSid: endpoint_sid,
+          preferenceId: preference_id,
+          value: receive_preference == true
+        }
+      )
+      raise_if_response_error(response.body)
+      response.body
+    rescue Common::Client::Errors::ClientError => e
+      remap_error(e)
+    end
+
+    # Send a push notification to a single device
+    #
+    # @endpoint_sid String    the registration id as returned from `register`
+    # @template_id String     id of the push notification content template
+    # @personalisation Hash   data map provided by sender to fill in specified template
+    #
+    # @return Hash            response object
+    #
+    def send_notification(endpoint_id, template_id, personalization = {})
+      response = perform(
+        :post,
+        SEND_PATH, {
+          endpointSid: endpoint_id,
+          templateSid: template_id,
+          personalization: personalization
+        }
+      )
+      raise_if_response_error(response.body)
+      response.body
+    rescue Common::Client::Errors::ClientError => e
+      remap_error(e)
+    end
+
+    private
+
+    # Raise an error if the service returned an error in the
+    # body of a 200 response
+    #
+    def raise_if_response_error(body)
+      raise VEText::ResponseError, body if body[:success] == false
+    end
+
+    def remap_error(e)
```

Comment:
Could the remapping happen in middleware? See the [SOAP middleware](https://github.com/department-of-veterans-affairs/vets-api/blob/master/lib/common/client/middleware/response/soap_parser.rb) for an example.

Reply:
Based on the same chat I don't think we actually need this. I'll be rewriting the implementation/tests to reflect what the VEText API is actually doing. https://dsva.slack.com/archives/C01CSM3EZGT/p1615409557008500

Reply:
Is it necessary to remap a `Common::Client::Errors::ClientError` to a custom `Common::Exceptions::BackendServiceException`? Just not sure what the precedent is for this.

Reply:
When I started this PR, I did some poking around other service integrations in this repo, and this seemed like best practice. Happy to be wrong about that though.

Reply:
Sounds good to me. I'll get with Alastair to figure out this middleware stuff.